Save and execute futures in a task queue

highker opened the pull request:
Upon calling wait(), save the forked thread and the current thread to a
task queue. A idling thread (which currently is single threaded) should
pick a ready task and run till there is nothing in the task queue.

jamesr66a reviewed (approved):
LGTM!

jamesr66a commented:
BTW does backwards() work through this?

apaszke commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -538,12 +538,28 @@ RegisterOperators reg({
+              pop(stack, inputs[i]);
+            }
+            for (int i = n_inputs - 1; i >= 0; --i) {
+              push(*copied_stack, inputs[i]);
```
`Stack` is just `std::vector<IValue>`, so why do we need to intermediate step? You could simply push it directly to `(*copied_stack)[i]`.

apaszke commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -538,12 +538,28 @@ RegisterOperators reg({
+            }
+
+            // TODO: (thread pool) move the task to a queue for threads to pick up
+            auto task = [=](std::shared_ptr<Stack> new_stack) {
```
Actually this could have been a `unique_ptr`, which would also be faster.

apaszke commented on test/test_jit.py:
```diff
@@ -9398,6 +9398,46 @@ def wait_script(x):
+
+        x1 = torch.rand(1000, 1000)
+        x2 = torch.rand(1000, 1000)
+        x3 = torch.rand(1000, 1000)
```
Those are huge arrays and will make the test slow. Can you make them smaller?

zdevito commented on aten/src/ATen/core/ivalue.h:
```diff
@@ -457,19 +459,17 @@ struct CAFFE2_API IValue final {
-  IValue get() const {
-    AT_ASSERT(ready);
-    return result;
+  IValue get() {
```
This patch is a good way of testing the multi-threaded perf before we add a cooperative thread pool. However, this call to `future.get()` is going to cause the caller to block if the future is not ready. We eventually want a cooperative threading model, where when this method is called by the interpreter, if the future is not ready, then the interpreter suspends itself, schedules a task to complete itself, and then returns a Future to caller of the interpreter that can be waited on to get the result of the interpreter. Calling `wait` on the future would pull the current thread into the group of worker threads until that Future is finished.

I think the next step is to implement this "interpreter suspend" behavior, even if we do not have a thread pool that can use it. That way the only remaining step is to integrate the threadpool.

highker replied:
Thanks for the comments. Shall we make a stack of `(InterpreterState*, Stack*)` pair so that when the current interpreter suspends (i.e., `wait()`), it pushes its own `(InterpreterState*, Stack*)` to the stack and then the newly forked `(InterpreterState*, Stack*)`? A simple paradigm could be keeping calling `run()` for the top element of the stack till it is empty.
We can chat in person if that works for you.

zdevito commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -538,12 +539,24 @@ RegisterOperators reg({
-            push(stack, Future(pop(stack)));
+            // Move inputs to a separate stack
+            auto copied_stack = torch::make_unique<Stack>(
+                stack.begin() + (stack.size() - n_inputs),
```
isn't this just: `stack.end() - n_inputs`?

Commit:
Save and execute future tasks in a task queue

Upon calling wait(), save the forked thread and the current thread to a
task queue. A idling thread (which currently is single threaded) should
pick a ready task and run till there is nothing in the task queue.

highker commented:
@zdevito @jamesr66a @apaszke thanks for the review.
The PR has been completely written towards the thread pool design.
Might be somewhat a bit different from what we have discussed.

zdevito reviewed (commented):
Let's take some care to get the threadpool and future objects to look close to the design that Ilia suggested. My understanding is that the pseudocode looks something like this (missing is any actual parallel implementation). What I liked about this design is how it keeps the work items and futures very simple.

```
API for objects:


// c10::intrusive_ptr<Future> is what gets passed around
struct Future {
  IValue value() {
    assert(completed_);
    return value;
  }
  IValue wait() {
    if(completed_)
      return value_;
    global_work_queue.workOnTasksUntilCompleted(this);
  }
  bool completed() {
    return completed_;
  }
  void markCompleted(IValue value) {
    value_ = value;;
    completed_ true;
    for(callback : callbacks) {
      callback();
    }
  }
  // callbacks should not do any actual work. they should just schedule more work.
  void addCallback(callback) {
    callbacks.push_back(callback)
  }
private:
  IValue value_; // when finished the value
  bool completed_; // is this future complete
  std::vector<std::function<void(void)>> callbacks;
};


struct ThreadPool {
  void schedule(std::function<void(void)> work) {
    tasks.push_back(work);
  }
  void workOnTasksUntilCompleted(c10::intrusive_ptr<Future> until) {
    while(!until.completed()) {
      auto task = tasks.pop();
      task();
    }
  }
private:
  std::vector<std::function<void(void)>> tasks;
};



Then an interpreter run will be something like:

InterpreterState() {
  future = Future::create();
}


Future run(Stack& stack) {
  try {
    <interpreter loop>
  } catch(Suspend & e) {
    e.future->addCallback({
      global_work_queue.schedule({
        // continue working on this job
        Stack stack;
        this->run(stack);
      });
    });
    return future;
  }
  future->markCompleted(Tuple::create(<top-of-stack>));
  return future;
}
return future;
```

highker commented:
Comments addressed; two main concerns:
- The ownership of forked state/stack. I don't have a good place to hold them.
- The current `run` is still `void` given `wait` is a blocking call.

zdevito commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -533,19 +533,48 @@ RegisterOperators reg({
+
+            // TODO: who/how to hold the lifecycle for copied_stack/copied_state?
+            // We currently add them to callback so they will be destroyed when future completes.
+            // Suppose we have two threads: T1 forks T2 and waits on T2;
```
You can move the stack by using a class rather than a lambda. A class with an operator() is still a valid std::function<void(void)>.
```
class InterpreterContinuation {
   InterpreterContinuation(InterpreterState state, Stack initial_stack)
  : state(std::move(state)), stack(std::move(initial_stack)) {}
  void operator()(void) {
     state.run(stack);
  }
private:
  Stack stack;
  InterpreterState state;
};
```

After the interpreter returns, whether it finished or not, the stack is no longer used. So once this Closure finishes it no longer needs Stack and it will get destroyed. If the Interpreter did not finish, then it must have created a callback on another future that would schedule the rest of the interpreter to run. That callback will capture a reference to InterpreterState, increasing its refcount. InterpreterState will destruct when there are no longer any callbacks referring to it.

zdevito reviewed (commented):
I got halfway through but I need to run for a bit. I will finish the review this afternoon.

zdevito commented on aten/src/ATen/core/ivalue.h:
```diff
@@ -441,21 +443,60 @@ struct CAFFE2_API IValue final {
+    if (completed()) {
+      return;
+    }
+    c10::global_work_queue.workOnTasksUntilCompleted(this);
```
follow up: for now, look at `Scope.h` which does a similar thing. We will have to add the functionality to intrusive_ptr.

zdevito commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -645,25 +645,47 @@ struct InterpreterStateImpl {
-    while(pc < last) {
+
+    // Check if we are recovered from a suspend
+    if (current_inst) {
```
This logic is not necessary, see below.

zdevito commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -645,25 +645,47 @@ struct InterpreterStateImpl {
+            push(stack, e.future->value());
+            return;
           }
+          registerOutputs(inst, stack);
```
You can avoid the custom logic on startup by simply re-running the suspending instruction on startup (i.e. leaving the PC unchanged).  The rule should be: on a suspend, wait pops its arguments off the stack, but pushes nothing before suspending. That way when the suspend is thrown, the stack is restored to what it was before the wait ran. Re-running the wait will then continue execution.
```
try {
          auto & inst = instructions[pc];
          loadTensorsFromRegisters(inst.inputs, stack);	      
          size_t new_pc = pc + 1 + inst.callback(stack)
          for(int i = inst.outputs.size - 1; i >= 0; i--) {
              int reg = get(inst.outputs,i);	           
              registers[reg] = pop(stack)
          }
        
} catch(Suspend& e) {
    // same as before, pc was never updated, so restarting the interpreter will rerun the wait instruction, which will load the completed value onto the stack.
}
```
There should be no need for special logic about current_inst.

zdevito commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -678,7 +700,13 @@ struct InterpreterStateImpl {
         }
     }
     current_pc = pc;
+    if (stack.size() > 0) {
```
The stack is not guaranteed to be empty on entry in the interpreter. You can figure out how many things are returned by the interpreter by looking at the number of outputs in the original graph, and saving that number in Code. Also, there may be more than 1 output. 

```
if(num_outputs == 1)
  future->markCompleted(stack.back());
else
  future->markCompleted(Tuple::create(peekSlice(stack,0, num_outputs))
```

zdevito commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -645,25 +645,47 @@ struct InterpreterStateImpl {
+          try {
+            new_pc += inst.callback(stack);
+          } catch (Suspend& e) {
+            JIT_ASSERT(!e.future->completed());
```
this will be a race condition when multithreaded, between throw and catch another thread _could_ complete this future. Instead it is always safe to call addCallback on a future: it will just immediately call the callback if it is already completed.

zdevito commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -645,25 +645,47 @@ struct InterpreterStateImpl {
+            // Finish what we are waiting on.
+            // This will schedule the current thread itself back into the queue.
+            // TODO: we should throw this back to the executor.
+            e.future->wait();
```
This wait should not be here. The continuation is already scheduled, and the call to `.wait()` on the interpreter's future will ensure that progress is continued to be made.

zdevito commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -697,12 +725,21 @@ struct InterpreterStateImpl {
   // that it was originally used for.
   size_t current_pc = 0;
+  Instruction* current_inst = nullptr;
+  c10::intrusive_ptr<Future> future = c10::make_intrusive<Future>();
```
We should avoid allocating the future along the synchronous pathway. Here is my suggestion. We have 2 entry points:

```
void run(Stack& stack) {
    bool suspended = runImpl(stack);
    if(suspended)) {
       future.wait();
      if(num_outputs == 1)
         stack.push(future.value())
      else
        <unpack the tuple, pushing each value onto the stack>
   }

}
intrusive_ptr<Future> runAsync(Stack& stack) {
    if(!future) {
       future = make_instrusive<Future>();
    }
   runImpl(stack);
   return future;
}
// private, replaces current run
bool runImpl(Stack& stack) {
    <do normal run>
    ...
   catch(Suspend& e) {
      if(!future)
         future = make_future(); // we need to ensure this is allocated regardless
      return true; //we suspended and did not push the results to the stack
   }
     ...
    // under the synchronous pathway future will only be allocated on a suspend, 
    // so guard the mark of completion for the case where we never allocate it.
    if(future) {
        <markCompletedLogic>
    }
}
```

zdevito reviewed (commented):
Progress! I added a bunch more comments about how to structure things to allow both synchronous and asynchronous execution, and to avoid the messy restart-logic in the interpreter by simply re-running the wait instruction.

zdevito commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -533,19 +533,33 @@ RegisterOperators reg({
+
+            push(stack, forked_interprester.getFuture());
+
+            c10::global_work_queue.schedule(continuation);
```
`std::move(continuation)`

zdevito reviewed (approved):
Looks good! I only have minor comments, mostly to avoid having to pass `parent` around.

highker commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -645,26 +647,45 @@ struct InterpreterStateImpl {
-  void run(Stack & stack) {
+
+ private:
+  bool runImpl(Stack& stack, InterpreterState& parent) {
```
Will address this + `clone()` in an upcoming PR
